File: android/src/main/kotlin/com/syncrhrohealth/bandpass_ble_connection/core/handler/response/ImuDataRspHandler.kt
package com.syncrhrohealth.bandpass_ble_connection.core.handler.response

import android.util.Log
import com.syncrhrohealth.bandpass_ble_connection.core.handler.DeviceHandler
import com.syncrhrohealth.bandpass_ble_connection.core.model.IMUData
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * IMU Data Response Handler
 *
 * Parses the raw payload (including the single-byte type) into IMUData, using
 * integer and fractional components for each sensor reading (as in Python reference).fileciteturn12file1
 *
 * Data format (Packet.data):
 * Byte 0      : Payload type tag (0x03 = IMU data)
 * Byte 1–2    : Count (big-endian format)
 * Byte 3–10   : Accelerometer X (int32 LE integer + int32 LE fractional parts)
 * Byte 11–18  : Accelerometer Y (int32 LE integer + int32 LE fractional parts)
 * Byte 19–26  : Accelerometer Z (int32 LE integer + int32 LE fractional parts)
 * Byte 27–34  : Gyroscope X (int32 LE integer + int32 LE fractional parts)
 * Byte 35–42  : Gyroscope Y (int32 LE integer + int32 LE fractional parts)
 * Byte 43–50  : Gyroscope Z (int32 LE integer + int32 LE fractional parts)
 * Byte 51     : ADC tag (0x02)
 * Byte 52–53  : ADC raw value (big-endian uint16)
 * Byte 54     : RTC tag (0x04)
 * Byte 55–58  : RTC seconds (big-endian uint32)
 * Byte 59–60  : RTC milliseconds (big-endian uint16)
 * Byte 61     : Optional terminator tag – discard if present
 */


private val UTC_TZ = TimeZone.getTimeZone("UTC")
private val DATE_FMT = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply {
    timeZone = UTC_TZ
}
private val TIME_FMT = SimpleDateFormat("HH:mm:ss.SSS", Locale.US).apply {
    timeZone = UTC_TZ
}

/** Device’s “RTC base” (seconds from 1970-01-01 to 2000-01-01) */
private const val RTC_BASE_OFFSET = 946_684_800L


object ImuDataRspHandler {
    private const val TAG = "ImuDataRspHandler"

    fun handle(data: ByteArray, handler: DeviceHandler) {
//        Log.d(TAG, "handle: len=${data.size}, bytes=${data.toList()}")
        try {
            // 0: payload type
            require(data[0] == 0x03.toByte()) { "Invalid payload type: ${data[0]}" }

            // 1–2: count (big-endian)
            val count = ((data[1].toInt() and 0xFF) shl 8) or
                    (data[2].toInt() and 0xFF)

            // wrap for little-endian multi-byte reads
            val bb = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)

            // Parse sensor value: 32-bit integer part + 32-bit fractional part (1e6 scale)
            fun parseSensor(offset: Int): Float {
                val intPart = bb.getInt(offset)
                val fracPart = bb.getInt(offset + 4)
                return intPart + fracPart / 1_000_000f
            }

            // Accelerometer (first sample)
            val accelX = parseSensor(3)
            val accelY = parseSensor(11)
            val accelZ = parseSensor(19)

            // Gyroscope (first sample)
            val gyroX = parseSensor(27)
            val gyroY = parseSensor(35)
            val gyroZ = parseSensor(43)

            // ADC: tag + big-endian uint16
            require(data[51] == 0x02.toByte()) { "Invalid ADC tag: ${data[51]}" }
            val adcRaw = ((data[52].toInt() and 0xFF) shl 8) or
                    (data[53].toInt() and 0xFF)

            // ─── RTC: tag, seconds (uint32 BE), milliseconds (uint16 BE) ────────────
            var dateStr: String = ""
            var timeStr: String = ""

            if (data[54] == 0x04.toByte() && data.size >= 61) {
                //   seconds 55-58 (big-endian)
                val secs = ((data[55].toLong() and 0xFF) shl 24) or
                        ((data[56].toLong() and 0xFF) shl 16) or
                        ((data[57].toLong() and 0xFF) shl 8) or
                        (data[58].toLong() and 0xFF)

                //   ms 59-60 (big-endian)
                val ms = ((data[59].toInt() and 0xFF) shl 8) or
                        (data[60].toInt() and 0xFF)

                //   Align device epoch (2000-01-01) with Unix epoch (1970-01-01)
                val epochMillis = (secs + (RTC_BASE_OFFSET - 946_684_800L)) * 1_000L + ms

                val date = Date(epochMillis)
                dateStr = DATE_FMT.format(date)            // yyyy-MM-dd (UTC)
                timeStr = TIME_FMT.format(date)            // HH:mm:ss.SSS (UTC)
            }

            val imuData = IMUData(
                count,
                accelX, accelY, accelZ,
                gyroX, gyroY, gyroZ,
                adcRaw,
                dateStr,
                timeStr
            )

            // Log.d(TAG, "imuData: $imuData")

            handler.callBack.onImuDataRsp(imuData, handler)
        } catch (e: Exception) {
            Log.e(TAG, "Parsing error", e)
        }
    }
}

File: ios/Classes/Core/Handler/Response/ImuDataRspHandler.swift
//
//  ImuDataRspHandler.swift
//  Runner
//
//  Created by MAC on 10/3/25.
//

import Foundation

/**
 * IMU Data Response Handler
 *
 * Parses the raw payload (including the single-byte type) into IMUData, using
 * integer and fractional components for each sensor reading (as in Python reference).
 *
 * Data format (Packet.data):
 * Byte 0      : Payload type tag (0x03 = IMU data)
 * Byte 1–2    : Count (big-endian format)
 * Byte 3–10   : Accelerometer X (int32 LE integer + int32 LE fractional parts)
 * Byte 11–18  : Accelerometer Y (int32 LE integer + int32 LE fractional parts)
 * Byte 19–26  : Accelerometer Z (int32 LE integer + int32 LE fractional parts)
 * Byte 27–34  : Gyroscope X (int32 LE integer + int32 LE fractional parts)
 * Byte 35–42  : Gyroscope Y (int32 LE integer + int32 LE fractional parts)
 * Byte 43–50  : Gyroscope Z (int32 LE integer + int32 LE fractional parts)
 * Byte 51     : ADC tag (0x02)
 * Byte 52–53  : ADC raw value (big-endian uint16)
 * Byte 54     : RTC tag (0x04)
 * Byte 55–58  : RTC seconds (big-endian uint32)
 * Byte 59–60  : RTC milliseconds (big-endian uint16)
 * Byte 61     : Optional terminator tag – discard if present
 */

private let UTC_TZ = TimeZone(identifier: "UTC")!
private let DATE_FMT: DateFormatter = {
    let fmt = DateFormatter()
    fmt.dateFormat = "yyyy-MM-dd"
    fmt.timeZone = UTC_TZ
    fmt.locale = Locale(identifier: "en_US_POSIX")
    return fmt
}()
private let TIME_FMT: DateFormatter = {
    let fmt = DateFormatter()
    fmt.dateFormat = "HH:mm:ss.SSS"
    fmt.timeZone = UTC_TZ
    fmt.locale = Locale(identifier: "en_US_POSIX")
    return fmt
}()

/** Device's "RTC base" (seconds from 1970-01-01 to 2000-01-01) */
private let RTC_BASE_OFFSET: Int64 = 946_684_800

struct ImuDataRspHandler {
    static func handle(data: Data, handler: DeviceHandler) {
        // print("ImuDataRspHandler - handle: len=\(data.count), bytes=\(data.toHexString())")
        
        guard data.count >= 51 else {
            print("ImuDataRspHandler - Error: Not enough data received (minimum 51 bytes)")
            return
        }
        
        do {
            // 0: payload type
            guard data[0] == 0x03 else {
                print("ImuDataRspHandler - Error: Invalid payload type: \(data[0])")
                return
            }
            
            // 1–2: count (big-endian)
            let count = Int((UInt16(data[1]) << 8) | UInt16(data[2]))
            
            // Parse sensor value: 32-bit integer part + 32-bit fractional part (1e6 scale)
            func parseSensor(offset: Int) -> Float {
                let intPartData = data.subdata(in: offset..<(offset + 4))
                let fracPartData = data.subdata(in: (offset + 4)..<(offset + 8))
                
                let intPart = intPartData.toInt32(byteOrder: .LittleEndian)
                let fracPart = fracPartData.toInt32(byteOrder: .LittleEndian)
                
                return Float(intPart) + Float(fracPart) / 1_000_000.0
            }
            
            // Accelerometer (first sample)
            let accelX = parseSensor(offset: 3)
            let accelY = parseSensor(offset: 11)
            let accelZ = parseSensor(offset: 19)
            
            // Gyroscope (first sample)
            let gyroX = parseSensor(offset: 27)
            let gyroY = parseSensor(offset: 35)
            let gyroZ = parseSensor(offset: 43)
            
            // ADC: tag + big-endian uint16
            guard data.count >= 54 else {
                print("ImuDataRspHandler - Error: Not enough data for ADC (minimum 54 bytes)")
                return
            }
            
            guard data[51] == 0x02 else {
                print("ImuDataRspHandler - Error: Invalid ADC tag: \(data[51])")
                return
            }
            
            let adcData = data.subdata(in: 52..<54)
            let adcRaw = Int(adcData.toUInt16(byteOrder: .BigEndian))
            
            // ─── RTC: tag, seconds (uint32 BE), milliseconds (uint16 BE) ────────────
            var dateStr: String = ""
            var timeStr: String = ""
            
            if data.count >= 61 && data[54] == 0x04 {
                // seconds 55-58 (big-endian)
                let secsData = data.subdata(in: 55..<59)
                let secs = Int64(secsData.toUInt32(byteOrder: .BigEndian))
                
                // ms 59-60 (big-endian)
                let msData = data.subdata(in: 59..<61)
                let ms = Int64(msData.toUInt16(byteOrder: .BigEndian))
                
                // Align device epoch (2000-01-01) with Unix epoch (1970-01-01)
                // Note: Matching Android implementation exactly
                let epochMillis = (secs + (RTC_BASE_OFFSET - 946_684_800)) * 1_000 + ms
                
                let date = Date(milliseconds: epochMillis)
                dateStr = DATE_FMT.string(from: date)            // yyyy-MM-dd (UTC)
                timeStr = TIME_FMT.string(from: date)            // HH:mm:ss.SSS (UTC)
            }
            
            let imuData = IMUData(
                count: count,
                accelX: accelX,
                accelY: accelY,
                accelZ: accelZ,
                gyroX: gyroX,
                gyroY: gyroY,
                gyroZ: gyroZ,
                adcRaw: adcRaw,
                date: dateStr,
                timeMs: timeStr
            )
            
            // print("ImuDataRspHandler - imuData: \(imuData)")
            
            handler.getCallBackToCentral()?.onImuDataRsp(imuData: imuData, handler: handler)
        } catch {
            print("ImuDataRspHandler - Error parsing data: \(error)")
        }
    }
}

